enhancements/node/lazy-image-pull.md

```diff
@@ -0,0 +1,266 @@
+* As a developer, I want to deploy applications packaged in large container images while minimizing the impact of image size on pod startup time.
+
+# Goals
+
+* Add support for drop-in config in [container storage config](https://github.com/containers/storage/blob/main/pkg/config/config.go)
+* Package upstream [Stargz Store plugin](https://github.com/containerd/stargz-snapshotter) for the installation on RHCOS
+* Enable lazy image pulling in Openshift using [Stargz Store plugin](https://github.com/containerd/stargz-snapshotter) with [zstd:chunked](https://github.com/containers/storage/pull/775) image format.
+* Support lazy image pulling only on the worker nodes (or subset of worker nodes).
+
+# Non-Goals
+
+* Enabling lazy image pulling on the control plane nodes.
+* Replacement of the standard OCI image format; `zstd:chunked` is a variant of zstd compressed layers in OCI, which is supported by podman and moby
+* Alterations to the image build process; developers can use standard OCI build tools, such as buildah to build and convert images to `zstd:chunked` format.
+* Support for lazy image pull format other than `zstd:chunked`
+
+## Proposal
+
+In order to use `zstd:chunked` images, we need to have [Stargz Store plugin](https://github.com/containerd/stargz-snapshotter/blob/aaa46a75dd97e401025f82630c9d3d4e41c9f670/docs/INSTALL.md#install-stargz-store-for-cri-opodman-with-systemd) installed on the node and with a configuration for container storage library. [ContainerRuntimeConfig](https://github.com/openshift/machine-config-operator/blob/release-4.16/docs/ContainerRuntimeConfigDesign.md) will be used to enable this feature in OCP.
+
+### Workflow Description
+
+* A cluster administrator will create or edit the [ContainerRuntimeConfig](https://github.com/openshift/machine-config-operator/blob/release-4.16/docs/ContainerRuntimeConfigDesign.md) object to enable lazy image pulling support by,
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+ name: lazy-image-pull
+spec:
+ machineConfigPoolSelector:
+   matchLabels:
+     pools.operator.machineconfiguration.openshift.io/worker: ""
+ containerRuntimeConfig:
+     lazyImagePull:
+        enabled: true
+```
+* Optionally, A list of supported types of the image format for lazy pulling can also be specified. At the time of writing this document, Openshift will only support `zstd:chunked` type image format. If the user doesn't explicitly specify the image format, the default value is set to `zstd:chunked`.
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+ name: lazy-image-pull
+spec:
+ machineConfigPoolSelector:
+   matchLabels:
+     pools.operator.machineconfiguration.openshift.io/worker: ""
+ containerRuntimeConfig:
+     lazyImagePull:
+        enabled: true
+        format:
+          - "zstd:chunked"
```

Comment:
Right now, none of this exists at a generic level: the on-demand layer plugin is provided with a an image reference and a layer digest, and that’s it. So if this had to be implemented _today_, that plugin would need to re-parse the image manifest, find the layer, and impose its own conditions on the layer’s metadata.

If we need to add a generic feature, that will need to be co-designed along with this enhancement.

Reply:
>How is this value consumed by the runtime? i.e. How can the runtime use this value for deciding whether it enable lazy pulling on the pulling image? Maybe this should be a mediatype + manifest annotation?

@ktock As of now, this values isn't directly used by the runtime. We are using this value to in MCO to prepare the configuration file for the `c/storage` library.

Reply:
I think we can introduce a field `additionallayerstore_target_mediatype` in storage.conf. This is layer mediatypes to enable to lookup of Additional Layer Store.

`containerRuntimeConfig.lazyImagePull.format` should also be a list of mediatype. Or, MCO can define aliases for them; e.g. `"zstd:chunked"` means `application/vnd.oci.image.layer.v1.tar+zstd`.

Reply:
My first instinct, from the runtime and c/storage perspective is to ask “why do we need that option at all?”.

I could, _maybe_, imagine that users might want to enable/disable lazy pulls based on the application (= on the registry namespace, on the K8S namespace, or as a Pod option). (To be clear, I don’t think we _need_ that for a first version at all.)

But why would users want to make a choice based on the format? Isn’t that ~transparent to the running systems, other than performance?

---

It might be necessary to have an option that governs “what format backends (snapshotters?) should be installed on nodes” — maybe not for the first version, but it does make sense to ensure that we can add that option later.

But that option should govern the behavior of MCO, c/storage doesn’t need to know. `storage.conf` only contains an `additionallayerstore` option with a list of mount points; how those mount points get set up and what are the backing FUSE servers is not a `storage.conf` concern AFAICS.

Reply:
Thanks for the review. Removed this option from the proposal.